docs/proposals/affinity.md

```diff
@@ -0,0 +1,423 @@
+rules are *hints* that the scheduler will try to satisfy but may not be able to.
+This is similar to the difference between
+[HardNodeAffinity and SoftNodeAffinity](https://github.com/kubernetes/kubernetes/pull/18261).
+
+Note that this proposal is only about *inter-pod* topological affinity and anti-affinity.
+There are other forms of topological affinity and anti-affinity. For example,
+you can use HardNodeAffinity (SoftNodeAffinity) to require (prefer)
+that a set of pods all be scheduled in some zone Z. SoftNodeAffinity and HardNodeAffinity
+are not capable of expressing inter-pod dependencies, and conversely
+the API we descibe here is not capable of expressing SoftNodeAffinity and HardNodeAffinity rules.
+For simplicity, we will use the terms "affinity" and "anti-affinity" to mean
+"inter-pod topological affinity" and "inter-pod topological anti-affinity" in the remainder
+of this document.
+
+## API
+
+We will add one field to `PodSpec`
+
+```go
+Affinity *Affinity  `json:"affinity,omitempty"`
+```
+
+The `Affinity` type is defined as follows
+
+```go
+// The weights of all of the "soft" fields are added per-node to find the
+// most preferred node(s)
+type Affinity struct {
+	// When there are multiple elements, the lists of nodes corresponding to each
+	// PodAffinityTerm are intersected, i.e. all terms must be satisfied.
+	HardPodAffinity     []PodAffinityTerm  `json:"hardPodAffinity,omityempty"`
+	// When there are multiple elements, the lists of nodes corresponding to each
+	// PodAffinityTerm are intersected, i.e. all terms must be satisfied.
+	HardPodAntiAffinity []PodAffinityTerm  `json:"hardPodAntiAffinity,omityempty"`
+	// SoftPodAffinity specifies which nodes are preferred among those that meet all
+    // of the hard scheduling requirements (resource request, HardPodAffinity, etc.).
+    // The node that is most preferred is the one with the greatest sum of weights,
+    // i.e. for each node that meets all of the hard scheduling requirements,
+    // compute a sum by iterating through the elements of SoftPodAffinity and adding
+    // "weight" to the sum if the corresponding PodAffinityTerm is satisfied for
+    // that node; the node(s) with the highest sum are the most preferred.
+	SoftPodAffinity     []SoftPodTerm  `json:"softPodAffinity,omityempty"`
+	// Same as SoftPodAffinity, but specifying anti-affinity.
+	SoftPodAntiAffinity []SoftPodTerm  `json:"softPodAntiAffinity,omityempty"`
+}
+
+type SoftPodTerm struct {
+    // weight is in the range 1-100
+    int weight  `json:"weight"`
+    PodAffinityTerm PodAffinityTerm  `json:"podAffinityTerm"`
+}
+
+type PodAffinityTerm struct {
+	LabelSelector *LabelSelector `json:"labelSelector,omitempty"`
+	// namespaces specifies which namespaces the LabelSelector applies to (matches against);
+	// nil list means "this pod's namespace," empty list means "all namespaces"
+	Namespaces []api.Namespace  `json:"namespaces,omitempty"`
```

Comment:
The json tag here should not be specified with omitempty, since nil and empty have different meanings.
Otherwise a non-nil empty []api.Namespace value of Namespaces will be ignored in json.Marshal().

Refer to https://golang.org/pkg/encoding/json/#Marshal
> Each exported struct field becomes a member of the object unless
```
- the field's tag is "-", or
- the field is empty and its tag specifies the "omitempty" option.
```
The empty values are false, 0, any nil pointer or interface value, and any array, slice, map, or string of length zero.